### ISSUE TITLE
My business notifications support

### ISSUE THREAD

**Issue by csga5000** [google/google-api-ruby-client#662]

**Comment by mattwhisenhunt** [comment]
It looks like the My Business API is not listed in the standard list of APIs. That's okay:
```sh
gem install google-api-client
generate-api gen lib --url=https://developers.google.com/my-business/samples/mybusiness_google_rest_v4.json
# It might complain about a missing api_names_out.yaml 
touch /full/path/to/lib/ruby/gems/*/gems/google-api-client-*/api_names_out.yaml
generate-api gen lib --url=https://developers.google.com/my-business/samples/mybusiness_google_rest_v4.json
```
The lib directory will be created in your current working directory. You may find it desirable to specify the `generated` directory in the google-api-client gem directory.
see
https://developers.google.com/my-business/samples/
https://github.com/google/google-api-ruby-client#generating-apis

**Comment by csga5000** [comment]
Well, even though it may not be listed you guys do support "my business", just it seems a feature is missing.  I can try the generator but isn't that what whom ever added my business to this repo did?

It's not listed because it's private and you have to register in a form to get access

**Comment by csga5000** [comment]
I agree, it is still a Google API, and should be included here.  With all features.  You already support v3, so I don't see why not add v4 and check that it included all relevant calls.  Especially when it's so easy to do so.

**Comment by hcancelik** [comment]
When can be expect v4 support for My Business API?

**Comment by csga5000** [comment]
Note, my business v3 was deleted: https://github.com/googleapis/google-api-ruby-client/commit/9d1a0e4b9e88d96373df48161e9c5e1035265ccd

To use my business without generating the API using their tool you would have to check out 0.23.7.  Though no version seems to work to me for my business (see https://github.com/googleapis/google-api-ruby-client/issues/712)

**Comment by csga5000** [comment]
@mattwhisenhunt Attempting your guide.

Added to my Gemfile
```
gem 'google-api-client', "~> 0.24.2"
```

Execute:
```
bundle install
generate-api gen lib --url=https://developers.google.com/my-business/samples/mybusiness_google_rest_v4.json
```

Errors: ```Thor is required. Please install the gem with development dependencies.```

So I execute: ```gem install --dev thor```

Try again, same error.

Any idea why?

**Comment by mattwhisenhunt** [comment]
I think the thor executable needs to be in your path.

**Comment by csga5000** [comment]
@mattwhisenhunt I thought to check that I could execute thor as well, so I had run:

```
ubuntu@ubuntu-xenial:~/myproj/api$ thor
Commands:
  thor help [COMMAND]  # Describe available commands or one specific command
  thor install NAME    # Install an optionally named Thor file into your system commands
  thor installed       # List the installed Thor modules and commands
  thor list [SEARCH]   # List the available thor commands (--substring means .*SEARCH)
  thor uninstall NAME  # Uninstall a named Thor module
  thor update NAME     # Update a Thor file from its original location
  thor version         # Show Thor version
```

Seems to be accessible to me?  (And these kinds of issues, as well as the fact that you can no-longer just checkout a gem when deploying to live and such, is exactly why these libraries should be added to this repo)

**Comment by csga5000** [comment]
It's in my path because I have rvm properly configured.

Also note, both generate-api and thor are in the same gemset:
```
ubuntu@ubuntu-xenial:~/myproject/api$ which thor
/home/ubuntu/.rvm/gems/ruby-2.4.1@myproject/bin/thor
ubuntu@ubuntu-xenial:~/myproject/api$ which generate-api
/home/ubuntu/.rvm/gems/ruby-2.4.1@myproject/bin/generate-api
```

**Comment by csga5000** [comment]
To clarify, what you should do for this is either as follows:
```
gem install google-api-client
gem install --dev thor
```

Or if you're using it in a gemfile and what not, your gemfile should have this in your gemfile
```
gem 'google-api-client', "~> 0.24.2"
group :development do
    gem 'thor'
end
```

## Either way, you then move on to here:
Run this:
```
generate-api gen lib --url=https://developers.google.com/my-business/samples/mybusiness_google_rest_v4.json
```

You'll probably get an error about failing to open a api_names_out.yaml file, copy the file path for that file, and run: ```touch /path-to-gems-copied-before/gems/google-api-client-0.24.2/api_names_out.yaml```

Now, rerun the command: ```generate-api gen lib --url=https://developers.google.com/my-business/samples/mybusiness_google_rest_v4.json```

Keep in mind, when deploying this to a live server, you'll need to ensure the generated api's are present there, and additionally handle for updates that may occur as it can't be versioned in your Gemfile like most dependencies.  Which @mattwhisenhunt is another reason this should be versioned in this project.

**Comment by andru1989** [comment]
Hi @csga5000, At this moment I need to use one of the Google My Business API V4 functionalities, like authentication, list accounts, locations, reviews, etc, because I couldn't do it with V3

Does `0.24.2` version supports this ? And does it have any documentation ?

Thanks in advance

**Comment by csga5000** [comment]
@lcpriest 
Well google no longer even lists v3 in their documentation: https://developers.google.com/my-business/reference/rest/
And that commit was autogenerated, so I assume it was removed due to deprecation.
The devs of this ruby library seem opposed to adding built in support because "it is not listed in the standard list of apis".  Why that matters, I'm not certain, the real puzzler is if that's the "rule", then why was there ever v3.

**Comment by dazuma** [comment]
Apologies for the uncertainty here. You can take this as the authoritative response from the internal Google team that now maintains this library.

This library covers the public APIs that are listed in the [Google API discovery service](https://developers.google.com/discovery/), but intentionally does not include libraries not in that list. There are various reasons why an API would not be in that list, and would not be covered by this library. For example, an API might not be generally available, but accessible only to certain enabled partners. That is the case for this API, Google My Business. We omit those from the library to avoid confusion, as well as to prevent the library from getting too large by including the long tail of APIs that are available only to a very small number of customers. We also might omit APIs that require certain specialized handling or access features not available in this library's mechanisms.

At this point, Google My Business V4 is not part of the standard distribution of this library. As earlier commenters have mentioned, you can use this library to generate your own client classes for the API. (And we just committed a fix for the `api_names_out.yaml` issue, so that should no longer be a problem in the next release.) As for why V3 _was_ present in this library earlier, unfortunately I don't know. It might have simply been a policy change in how the My Business team decided to handle clients.

Finally, apologies for the delayed response here. My team only recently inherited maintenance of this library, and we're just now trying to catch up on the backlog of questions and requests. The good news is that we do now have a dedicated team working on it (including actual experienced Ruby developers! 🎉) so we should be able to provide better support moving forward.

**Comment by skarmakar** [comment]
@csga5000 @dazuma I know this is a pretty old issue. But I do not find any updated working doc/upgradation of this library. The process mentioned here does not work anymore. Any kind of help about how to use google my business APIs using this library would be a great help. Thanks!

**Comment by fraasssee** [comment]
@skarmakar Not sure if you still have issues but I managed to find the right URL, https://developers.google.com/static/my-business/samples/mybusiness_google_rest_v4p9.json, that allowed me to generate the lib and eventually getting it to work.

**Comment by skarmakar** [comment]
@fraasssee awesome! I shall check this. Thanks for the help!
